sql/codegen_xgboost.go

```diff
@@ -14,41 +14,109 @@
-	Maximize      bool `json:"maximize,omitempty"`
+	NumBoostRound       int
+	Maximize            bool
+	EarlyStoppingRounds int
 }
 
 type xgbFiller struct {
-	IsTrain              bool
-	TrainingDatasetSQL   string
-	ValidationDatasetSQL string
-	TrainCfg             *xgbTrainConfig
-	Features             []*featureMeta
-	Label                *featureMeta
-	Save                 string
-	ParamsCfgJSON        string
-	TrainCfgJSON         string
-	*connectionConfig
+	Estimator
+	xgbTrainConfig
+	Save          string
+	ParamsCfgJSON string
+}
+
+func resolveTrainCfg(attrs map[string]*attribute) *xgbTrainConfig {
+	return &xgbTrainConfig{
+		NumBoostRound:       getIntAttr(attrs, "train.num_boost_round", 10),
+		Maximize:            getBoolAttr(attrs, "train.maximize", false, false),
+		EarlyStoppingRounds: getIntAttr(attrs, "train.early_stopping_rounds", -1),
+	}
 }
 
-func newXGBFiller(pr *extendedSelect, ds *trainAndValDataset, fts fieldTypes, db *DB) (*xgbFiller, error) {
+func resolveParamsCfg(attrs map[string]*attribute) (map[string]interface{}, error) {
+	// extract the attributes without any prefix as the XGBoost Parmaeters
+	params := make(map[string]interface{})
 	var err error
+	for k, v := range attrs {
+		if !strings.Contains(k, ".") {
+			var vStr string
+			var ok bool
+			if vStr, ok = v.Value.(string); !ok {
+				return nil, fmt.Errorf("convert params %s to string failed, %v", vStr, err)
+			}
+			if govalidator.IsFloat(vStr) {
+				floatVal, err := strconv.ParseFloat(vStr, 16)
+				if err != nil {
+					return nil, fmt.Errorf("convert params %s to float32 failed, %v", vStr, err)
+				}
+				params[k] = floatVal
+			} else if govalidator.IsInt(vStr) {
+				if params[k], err = strconv.ParseInt(vStr, 0, 32); err != nil {
+					return nil, fmt.Errorf("convert params %s to int32 failed, %v", vStr, err)
+				}
+			} else if govalidator.IsASCII(vStr) {
+				params[k] = vStr
+			} else {
+				return nil, fmt.Errorf("unsupported params type: %s", vStr)
+			}
+		}
+	}
+	return params, nil
+}
+
+func resolveObjective(pr *extendedSelect) (string, error) {
+	estimatorParts := strings.Split(pr.estimator, ".")
+	if len(estimatorParts) != 3 {
+		return "", fmt.Errorf("XGBoost Estimator should be xgboost.first_part.second_part")
+	}
+	return strings.Join(estimatorParts[1:], ":"), nil
+}
+
+func newXGBFiller(pr *extendedSelect, ds *trainAndValDataset, db *DB) (*xgbFiller, error) {
+	attrs, err := resolveAttribute(&pr.trainAttrs)
+	if err != nil {
+		return nil, err
+	}
 	training, validation := trainingAndValidationDataset(pr, ds)
 	r := &xgbFiller{
-		IsTrain:              pr.train,
-		TrainingDatasetSQL:   training,
-		ValidationDatasetSQL: validation,
-		Save:                 pr.save,
+		Estimator: Estimator{
+			IsTrain:              pr.train,
+			TrainingDatasetSQL:   training,
+			ValidationDatasetSQL: validation,
+		},
+		xgbTrainConfig: *resolveTrainCfg(attrs),
+		Save:           pr.save,
+	}
+
+	params, err := resolveParamsCfg(attrs)
```

Comment:
`resolveParamsCfg` must run after `resolveTrainCfg` to resolve all other attributes, maybe add some comment here as hints.

Reply:
`resolveParamsCfg ` only resolve the keys without prefix, maybe the order is fine?